codegenerator/cli/src/hbs_templating/codegen_templates.rs

```diff
@@ -342,12 +343,72 @@ pub struct EventTemplate {
     pub convert_hyper_sync_event_args_code: String,
     pub data_type: String,
     pub data_schema_code: String,
+    pub get_topic_selection_code: String,
+    pub event_filter_type: String,
 }
 
 impl EventTemplate {
     const DECODE_HYPER_FUEL_DATA_CODE: &'static str =
         "(_) => Js.Exn.raiseError(\"HyperFuel decoder not implemented\")";
 
+    const GET_TOPIC_SELECTION_CODE_STUB: &'static str =
+        "_ => [LogSelection.makeTopicSelection(~topic0=[sighash->EvmTypes.Hex.\
+         fromStringUnsafe])->Utils.unwrapResultExn]";
+
+    const EVENT_FILTER_TYPE_STUB: &'static str = "{}";
+
+    pub fn generate_event_filter_type(params: &Vec<EventParam>) -> String {
+        let field_rows = params
+            .iter()
+            .filter(|param| param.indexed)
+            .map(|param| {
+                format!(
+                    "@as(\"{}\") {}?: SingleOrMultiple.t<{}>",
+                    param.name,
+                    RescriptRecordField::to_valid_res_name(&param.name),
+                    abi_to_rescript_type(&param.into())
+                )
+            })
+            .collect::<Vec<_>>()
+            .join(", ");
+
+        format!("{{ {field_rows} }}")
+    }
+
+    pub fn generate_get_topic_selection_code(params: &Vec<EventParam>) -> String {
+        let indexed_params = params.iter().filter(|param| param.indexed);
+
+        //Prefixed with underscore for cases where it is not used to avoid compiler warnings
+        let event_filter_arg = "_eventFilter";
+
+        let topic_filter_calls = indexed_params
+            .enumerate()
+            .map(|(i, param)| {
+                let param = EthereumEventParam::from(param);
+                let topic_number = i + 1;
+                let param_name = RescriptRecordField::to_valid_res_name(param.name);
+                let topic_encoder = param.get_topic_encoder();
+                let nested_type_flags = match param.get_nested_type_depth() {
+                    depth if depth > 0 => format!("(~nestedArrayDepth={depth})"),
+                    _ => "".to_string(),
+                };
+                format!(
+                    "~topic{topic_number}=?{event_filter_arg}.{param_name}->Belt.Option.\
+                     map(topicFilters => \
+                     topicFilters->SingleOrMultiple.normalizeOrThrow{nested_type_flags}->Belt.\
+                     Array.map({topic_encoder})), "
+                )
+            })
+            .collect::<String>();
+
+        format!(
+            "(eventFilters) => \
+             eventFilters->SingleOrMultiple.normalizeOrThrow->Belt.Array.map({event_filter_arg} \
+             => LogSelection.makeTopicSelection(~topic0=[sighash->EvmTypes.Hex.fromStringUnsafe], \
+             {topic_filter_calls})->Utils.unwrapResultExn)"
+        )
+    }
```

Comment:
This is completely covered in test_codegen test but I haven't written any unit tests on this directly in the rust codebase.